File: ios/Sources/CompanionCore/MessageAttachments.swift
import Foundation

/// One attachment waiting in the mobile composer. The bytes are app-owned:
/// picker URLs are copied before this value is created, so a later send never
/// depends on a security-scoped URL still being alive.
public struct PendingMessageAttachment: Identifiable, Hashable, Sendable {
    public enum Kind: Hashable, Sendable {
        case image
        case file
    }

    public let id: UUID
    public let data: Data
    public let name: String
    public let mime: String
    public let kind: Kind

    public init(
        id: UUID = UUID(),
        data: Data,
        name: String,
        mime: String,
        kind: Kind
    ) {
        self.id = id
        self.data = data
        self.name = name
        self.mime = mime
        self.kind = kind
    }

    public var bytes: Int { data.count }
}

public enum AttachmentPolicyError: Error, LocalizedError, Equatable, Sendable {
    case tooManyItems
    case totalTooLarge
    case invalidName
    case unsupportedType(String)
    case itemTooLarge(name: String, limitMB: Int)

    public var errorDescription: String? {
        switch self {
        case .tooManyItems:
            return "Attach up to 4 items at a time."
        case .totalTooLarge:
            return "Those attachments are larger than 50 MB together."
        case .invalidName:
            return "That file doesn't have a valid filename."
        case let .unsupportedType(name):
            return "\(name) isn't a supported file. Try PDF, text, Word, Excel, or PowerPoint."
        case let .itemTooLarge(name, limitMB):
            return "\(name) is larger than \(limitMB) MB."
        }
    }
}

/// The same limits apply to the in-app composer and the Share extension.
/// Keeping the policy in CompanionCore prevents either entry point from
/// accepting an attachment that the authenticated upload route will reject.
public enum AttachmentPolicy {
    public static let maximumItems = 4
    public static let maximumTotalBytes = 50 * 1_024 * 1_024
    public static let maximumImageBytes = 10 * 1_024 * 1_024
    public static let maximumFileBytes = 25 * 1_024 * 1_024

    public static let imageMIMETypes: Set<String> = [
        "image/png", "image/jpeg", "image/gif", "image/webp",
    ]

    public static let documentMIMETypes: Set<String> = [
        "text/plain", "text/markdown", "text/csv", "text/tab-separated-values",
        "application/json", "application/pdf", "application/rtf", "text/rtf",
        "application/msword",
        "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
        "application/vnd.ms-excel",
        "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
        "application/vnd.ms-powerpoint",
        "application/vnd.openxmlformats-officedocument.presentationml.presentation",
        "application/vnd.oasis.opendocument.text",
        "application/vnd.oasis.opendocument.spreadsheet",
        "application/vnd.oasis.opendocument.presentation",
    ]

    public static func normalizedMIME(_ value: String) -> String {
        value.split(separator: ";", maxSplits: 1, omittingEmptySubsequences: true)
            .first.map(String.init)?
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .lowercased() ?? ""
    }

    public static func kind(forMIME value: String) -> PendingMessageAttachment.Kind? {
        let mime = normalizedMIME(value)
        if imageMIMETypes.contains(mime) { return .image }
        if documentMIMETypes.contains(mime) { return .file }
        return nil
    }

    public static func validate(_ attachments: [PendingMessageAttachment]) throws {
        guard attachments.count <= maximumItems else { throw AttachmentPolicyError.tooManyItems }
        guard attachments.reduce(0, { $0 + $1.data.count }) <= maximumTotalBytes else {
            throw AttachmentPolicyError.totalTooLarge
        }

        for attachment in attachments {
            let name = attachment.name.trimmingCharacters(in: .whitespacesAndNewlines)
            guard validDisplayName(name) else { throw AttachmentPolicyError.invalidName }
            let mime = normalizedMIME(attachment.mime)
            switch attachment.kind {
            case .image:
                guard imageMIMETypes.contains(mime) else {
                    throw AttachmentPolicyError.unsupportedType(name)
                }
                guard attachment.data.count <= maximumImageBytes else {
                    throw AttachmentPolicyError.itemTooLarge(name: name, limitMB: 10)
                }
            case .file:
                guard documentMIMETypes.contains(mime) else {
                    throw AttachmentPolicyError.unsupportedType(name)
                }
                guard attachment.data.count <= maximumFileBytes else {
                    throw AttachmentPolicyError.itemTooLarge(name: name, limitMB: 25)
                }
            }
        }
    }

    static func validMIME(_ value: String) -> Bool {
        let mime = normalizedMIME(value)
        guard !mime.isEmpty, mime.utf8.count <= 127, mime.contains("/") else { return false }
        return mime.utf8.allSatisfy { byte in
            (48...57).contains(byte) || (65...90).contains(byte) || (97...122).contains(byte)
                || [33, 35, 36, 38, 43, 45, 46, 47, 94, 95].contains(byte)
        }
    }

    private static func validDisplayName(_ value: String) -> Bool {
        !value.isEmpty && value.utf8.count <= 255
            && !value.contains("/") && !value.contains("\\")
            && !value.unicodeScalars.contains(where: CharacterSet.controlCharacters.contains)
    }
}

/// What tapping a Markdown link in a message is allowed to do. Web links go
/// to the system. Absolute desktop paths go back through the authenticated
/// companion file route. Relative paths are resolved by that route against
/// the originating conversation's workspace. Custom schemes do nothing.
public enum LocalMessageLink: Equatable, Sendable {
    case web(URL)
    case desktopFile(path: String)

    public static func resolve(_ rawValue: String) -> LocalMessageLink? {
        let value = rawValue.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !value.isEmpty, value.utf8.count <= 8_192,
              !value.unicodeScalars.contains(where: CharacterSet.controlCharacters.contains)
        else { return nil }

        if isWindowsAbsolutePath(value) || isBackslashUNCPath(value) {
            return .desktopFile(path: value)
        }
        // In Markdown, //host/path is a protocol-relative web link. Keep it
        // out of the UNC path branch; file://host/path remains the explicit
        // spelling for a network share.
        if value.hasPrefix("//") {
            guard let components = URLComponents(string: "https:\(value)"),
                  let url = components.url,
                  components.host?.isEmpty == false
            else { return nil }
            return .web(url)
        }
        if value.hasPrefix("/") { return .desktopFile(path: value) }

        if let schemeRange = value.range(
            of: #"^[A-Za-z][A-Za-z0-9+.-]*:"#,
            options: .regularExpression
        ) {
            guard schemeRange.lowerBound == value.startIndex,
                  let components = URLComponents(string: value),
                  let scheme = components.scheme?.lowercased()
            else { return nil }
            if scheme == "http" || scheme == "https" {
                guard let url = components.url, components.host?.isEmpty == false else { return nil }
                return .web(url)
            }
            if scheme == "file" {
                guard components.query == nil, components.fragment == nil else { return nil }
                guard var path = components.percentEncodedPath.removingPercentEncoding else { return nil }
                if path.hasPrefix("/"), isWindowsAbsolutePath(String(path.dropFirst())) {
                    path.removeFirst()
                }
                if let host = components.host, !host.isEmpty, host.lowercased() != "localhost" {
                    let sharePath = path.replacingOccurrences(of: "/", with: "\\")
                    path = "\\\\\(host)\(sharePath)"
                }
                guard path.hasPrefix("/") || isWindowsAbsolutePath(path) || isBackslashUNCPath(path) else {
                    return nil
                }
                return .desktopFile(path: path)
            }
            return nil
        }

        // Query strings and fragments belong to the Markdown target, not the
        // local filename. Decode the path the same way as the scoped server
        // matcher, while refusing query-only/fragment-only links.
        let delimiter = value.firstIndex(where: { $0 == "?" || $0 == "#" }) ?? value.endIndex
        let pathPart = String(value[..<delimiter])
        guard !pathPart.isEmpty, let path = pathPart.removingPercentEncoding, !path.isEmpty else { return nil }
        return .desktopFile(path: path)
    }

    public static func resolve(_ url: URL) -> LocalMessageLink? {
        let value = url.absoluteString
        if let decoded = value.removingPercentEncoding,
           isWindowsAbsolutePath(decoded) || isBackslashUNCPath(decoded) {
            return .desktopFile(path: decoded)
        }
        return resolve(value)
    }

    private static func isWindowsAbsolutePath(_ value: String) -> Bool {
        let bytes = Array(value.utf8)
        guard bytes.count >= 3 else { return false }
        let letter = (65...90).contains(bytes[0]) || (97...122).contains(bytes[0])
        return letter && bytes[1] == 58 && (bytes[2] == 47 || bytes[2] == 92)
    }

    private static func isBackslashUNCPath(_ value: String) -> Bool {
        value.hasPrefix("\\\\")
    }
}

/// Bytes returned by the authenticated file route. `localURL` is nil on the
/// low-level client and populated by Session after it safely writes a preview
/// into the app's temporary directory.
public struct DownloadedFile: Sendable {
    public let data: Data
    public let filename: String
    public let contentType: String
    public let localURL: URL?

    public init(data: Data, filename: String, contentType: String, localURL: URL? = nil) {
        self.data = data
        self.filename = filename
        self.contentType = contentType
        self.localURL = localURL
    }

    public func stored(at url: URL) -> DownloadedFile {
        DownloadedFile(data: data, filename: filename, contentType: contentType, localURL: url)
    }
}

File: android/core/src/main/kotlin/com/openmausbot/companion/core/MessageAttachments.kt
package com.openmausbot.companion.core

import java.net.URI
import java.net.URLDecoder
import java.util.UUID
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.CoroutineStart
import kotlinx.coroutines.Deferred
import kotlinx.coroutines.async

/**
 * One attachment waiting in the composer — the port of
 * `ios/Sources/CompanionCore/MessageAttachments.swift`.
 *
 * The bytes are app-owned: picker URIs are copied before this value is created,
 * so a later send never depends on a content provider still granting access.
 */
class PendingMessageAttachment(
    val id: String = UUID.randomUUID().toString(),
    val data: ByteArray,
    val name: String,
    val mime: String,
    val kind: Kind,
) {
    enum class Kind { IMAGE, FILE }

    val bytes: Int get() = data.size
}

class AttachmentPolicyException(message: String) : Exception(message)

/**
 * The same limits apply to the in-app composer and the Share target. Keeping
 * the policy in `:core` prevents either entry point from accepting an
 * attachment that the authenticated upload route will reject.
 */
object AttachmentPolicy {
    const val MAXIMUM_ITEMS: Int = 4
    const val MAXIMUM_TOTAL_BYTES: Int = 50 * 1_024 * 1_024
    const val MAXIMUM_IMAGE_BYTES: Int = 10 * 1_024 * 1_024
    const val MAXIMUM_FILE_BYTES: Int = 25 * 1_024 * 1_024

    val IMAGE_MIME_TYPES: Set<String> = setOf("image/png", "image/jpeg", "image/gif", "image/webp")

    val DOCUMENT_MIME_TYPES: Set<String> = setOf(
        "text/plain", "text/markdown", "text/csv", "text/tab-separated-values",
        "application/json", "application/pdf", "application/rtf", "text/rtf",
        "application/msword",
        "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
        "application/vnd.ms-excel",
        "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
        "application/vnd.ms-powerpoint",
        "application/vnd.openxmlformats-officedocument.presentationml.presentation",
        "application/vnd.oasis.opendocument.text",
        "application/vnd.oasis.opendocument.spreadsheet",
        "application/vnd.oasis.opendocument.presentation",
    )

    const val TOO_MANY_ITEMS: String = "Attach up to 4 items at a time."
    const val TOTAL_TOO_LARGE: String = "Those attachments are larger than 50 MB together."
    const val INVALID_NAME: String = "That file doesn't have a valid filename."

    fun unsupportedType(name: String): String =
        "$name isn't a supported file. Try PDF, text, Word, Excel, or PowerPoint."

    fun itemTooLarge(name: String, limitMB: Int): String = "$name is larger than $limitMB MB."

    /** `type/subtype` only, lowercased — a `; charset=` parameter is not part of the type. */
    fun normalizedMime(value: String): String =
        value.substringBefore(';').trim().lowercase()

    fun kindForMime(value: String): PendingMessageAttachment.Kind? {
        val mime = normalizedMime(value)
        if (mime in IMAGE_MIME_TYPES) return PendingMessageAttachment.Kind.IMAGE
        if (mime in DOCUMENT_MIME_TYPES) return PendingMessageAttachment.Kind.FILE
        return null
    }

    /** Throws [AttachmentPolicyException] with the sentence the composer shows. */
    fun validate(attachments: List<PendingMessageAttachment>) {
        if (attachments.size > MAXIMUM_ITEMS) throw AttachmentPolicyException(TOO_MANY_ITEMS)
        if (attachments.sumOf { it.data.size.toLong() } > MAXIMUM_TOTAL_BYTES) {
            throw AttachmentPolicyException(TOTAL_TOO_LARGE)
        }
        for (attachment in attachments) {
            val name = attachment.name.trim()
            if (!validDisplayName(name)) throw AttachmentPolicyException(INVALID_NAME)
            val mime = normalizedMime(attachment.mime)
            when (attachment.kind) {
                PendingMessageAttachment.Kind.IMAGE -> {
                    if (mime !in IMAGE_MIME_TYPES) throw AttachmentPolicyException(unsupportedType(name))
                    if (attachment.data.size > MAXIMUM_IMAGE_BYTES) {
                        throw AttachmentPolicyException(itemTooLarge(name, 10))
                    }
                }
                PendingMessageAttachment.Kind.FILE -> {
                    if (mime !in DOCUMENT_MIME_TYPES) throw AttachmentPolicyException(unsupportedType(name))
                    if (attachment.data.size > MAXIMUM_FILE_BYTES) {
                        throw AttachmentPolicyException(itemTooLarge(name, 25))
                    }
                }
            }
        }
    }

    /** A syntactically plausible media type, so a server header cannot smuggle anything odd. */
    fun validMime(value: String): Boolean {
        val mime = normalizedMime(value)
        if (mime.isEmpty() || mime.toByteArray().size > 127 || '/' !in mime) return false
        return mime.all { it in '0'..'9' || it in 'a'..'z' || it in 'A'..'Z' || it in "!#$&+-./^_" }
    }

    private fun validDisplayName(value: String): Boolean =
        value.isNotEmpty() && value.toByteArray().size <= 255 &&
            '/' !in value && '\\' !in value && value.none(Char::isISOControl)
}

/**
 * What tapping a Markdown link in a message is allowed to do. Web links go to
 * the system. Absolute desktop paths go back through the authenticated
 * companion file route. Relative paths are resolved by that route against the
 * originating conversation's workspace. Custom schemes do nothing.
 */
sealed class LocalMessageLink {
    data class Web(val url: String) : LocalMessageLink()
    data class DesktopFile(val path: String) : LocalMessageLink()

    companion object {
        fun resolve(rawValue: String): LocalMessageLink? {
            val value = rawValue.trim()
            if (value.isEmpty() || value.toByteArray().size > 8_192 || value.any(Char::isISOControl)) return null
            if (isWindowsAbsolutePath(value) || isBackslashUncPath(value)) return DesktopFile(value)
            // //host/path is a protocol-relative web link in Markdown. A UNC
            // share must use backslashes or the explicit file://host spelling.
            if (value.startsWith("//")) {
                val uri = runCatching { URI("https:$value") }.getOrNull() ?: return null
                if (uri.host.isNullOrEmpty()) return null
                return Web(uri.toString())
            }
            if (value.startsWith("/")) return DesktopFile(value)
            if (SCHEME.containsMatchIn(value)) {
                val uri = runCatching { URI(value) }.getOrNull() ?: return null
                val scheme = uri.scheme?.lowercase() ?: return null
                if (scheme == "http" || scheme == "https") {
                    if (uri.host.isNullOrEmpty()) return null
                    return Web(value)
                }
                if (scheme == "file") {
                    if (uri.rawQuery != null || uri.rawFragment != null) return null
                    var path = uri.rawPath?.let(::decodeOrNull) ?: return null
                    if (path.startsWith("/") && isWindowsAbsolutePath(path.drop(1))) path = path.drop(1)
                    val host = uri.host
                    if (!host.isNullOrEmpty() && host.lowercase() != "localhost") {
                        path = "\\\\$host${path.replace('/', '\\')}"
                    }
                    if (!(path.startsWith("/") || isWindowsAbsolutePath(path) || isBackslashUncPath(path))) {
                        return null
                    }
                    return DesktopFile(path)
                }
                return null
            }

            val delimiter = listOf(value.indexOf('?'), value.indexOf('#'))
                .filter { it >= 0 }
                .minOrNull() ?: value.length
            if (delimiter == 0) return null
            val path = decodeOrNull(value.substring(0, delimiter)) ?: return null
            return path.takeIf(String::isNotEmpty)?.let(::DesktopFile)
        }

        private val SCHEME = Regex("^[A-Za-z][A-Za-z0-9+.-]*:")

        private fun decodeOrNull(value: String): String? =
            runCatching { URLDecoder.decode(value.replace("+", "%2B"), "UTF-8") }.getOrNull()

        private fun isWindowsAbsolutePath(value: String): Boolean =
            value.length >= 3 && value[0].isLetter() && value[0].code < 128 && value[1] == ':' &&
                (value[2] == '/' || value[2] == '\\')

        private fun isBackslashUncPath(value: String): Boolean = value.startsWith("\\\\")
    }
}

/** A display/storage filename bounded by UTF-8 bytes without splitting a surrogate pair. */
internal fun sanitisePortableFilename(
    source: String,
    fallback: String,
    maximumUTF8Bytes: Int = 180,
): String {
    val basename = source.split('/', '\\').lastOrNull { it.isNotEmpty() } ?: fallback
    val output = StringBuilder()
    var bytes = 0
    var index = 0
    while (index < basename.length) {
        val character = basename[index]
        val (rawCodePoint, consumed) = when {
            Character.isHighSurrogate(character) &&
                index + 1 < basename.length && Character.isLowSurrogate(basename[index + 1]) ->
                Character.toCodePoint(character, basename[index + 1]) to 2
            Character.isHighSurrogate(character) || Character.isLowSurrogate(character) -> ' '.code to 1
            else -> character.code to 1
        }
        val codePoint = if (
            Character.isISOControl(rawCodePoint) ||
            rawCodePoint in 0x202A..0x202E ||
            rawCodePoint in 0x2066..0x2069
        ) {
            ' '.code
        } else {
            rawCodePoint
        }
        val piece = String(Character.toChars(codePoint))
        val pieceBytes = piece.toByteArray(Charsets.UTF_8).size
        if (bytes + pieceBytes > maximumUTF8Bytes) break
        output.append(piece)
        bytes += pieceBytes
        index += consumed
    }
    val cleaned = output.toString().trim()
    return cleaned.takeUnless { it.isEmpty() || it == "." || it == ".." } ?: fallback
}

/** Bytes returned by the authenticated file route, with the sanitised name and type to show them under. */
class DownloadedFile(
    val data: ByteArray,
    val filename: String,
    val contentType: String,
)

/** One authenticated attachment identity; connection id prevents cross-computer reuse. */
internal data class AttachmentDownloadKey(
    val connectionId: String,
    val threadId: String,
    val messageId: String,
    val path: String,
)

/**
 * Small session-owned LRU for transcript attachments. It also owns each active
 * request so two rows asking for the same image await one network transfer.
 */
internal class AttachmentDownloadCache(
    private val scope: CoroutineScope,
    private val maximumEntries: Int = 12,
    private val maximumBytes: Long = 50L * 1_024L * 1_024L,
) {
    private class Pending(val token: Any, var retainResult: Boolean) {
        lateinit var deferred: Deferred<DownloadedFile>
        var waiters: Int = 0
    }
    private sealed interface Lookup {
        data class Cached(val file: DownloadedFile) : Lookup
        data class PendingRequest(val pending: Pending, val created: Boolean) : Lookup
    }

    private val lock = Any()
    private val entries = LinkedHashMap<AttachmentDownloadKey, DownloadedFile>()
    private val inFlight = mutableMapOf<AttachmentDownloadKey, Pending>()
    private var storedBytes = 0L

    suspend fun getOrLoad(
        key: AttachmentDownloadKey,
        loader: suspend () -> DownloadedFile,
    ): DownloadedFile = getOrLoad(key, retainResult = true, loader = loader)

    suspend fun getOrLoad(
        key: AttachmentDownloadKey,
        retainResult: Boolean,
        loader: suspend () -> DownloadedFile,
    ): DownloadedFile {
        val lookup = synchronized(lock) {
            if (retainResult) {
                entries.remove(key)?.let { cached ->
                    entries[key] = cached
                    return@synchronized Lookup.Cached(cached)
                }
            }
            inFlight[key]?.let {
                it.waiters += 1
                it.retainResult = it.retainResult || retainResult
                return@synchronized Lookup.PendingRequest(it, created = false)
            }

            val token = Any()
            val pending = Pending(token, retainResult).apply { waiters = 1 }
            pending.deferred = scope.async(start = CoroutineStart.LAZY) {
                try {
                    val file = loader()
                    synchronized(lock) {
                        if (inFlight[key]?.token === token) {
                            inFlight.remove(key)
                            if (pending.retainResult) storeLocked(key, file)
                        }
                    }
                    file
                } finally {
                    synchronized(lock) {
                        if (inFlight[key]?.token === token) inFlight.remove(key)
                    }
                }
            }
            inFlight[key] = pending
            Lookup.PendingRequest(pending, created = true)
        }

        return when (lookup) {
            is Lookup.Cached -> lookup.file
            is Lookup.PendingRequest -> {
                if (lookup.created) lookup.pending.deferred.start()
                try {
                    lookup.pending.deferred.await()
                } finally {
                    releaseWaiter(key, lookup.pending)
                }
            }
        }
    }

    /** A computer switch or sign-out invalidates bytes and active transfers together. */
    fun clear() {
        val jobs = synchronized(lock) {
            entries.clear()
            storedBytes = 0
            inFlight.values.map { it.deferred }.also { inFlight.clear() }
        }
        jobs.forEach { it.cancel() }
    }

    private fun storeLocked(key: AttachmentDownloadKey, file: DownloadedFile) {
        val bytes = file.data.size.toLong()
        if (maximumEntries <= 0 || bytes > maximumBytes) return
        entries.remove(key)?.let { storedBytes -= it.data.size.toLong() }
        while (entries.isNotEmpty() && (entries.size >= maximumEntries || storedBytes + bytes > maximumBytes)) {
            val eldest = entries.entries.first()
            entries.remove(eldest.key)
            storedBytes -= eldest.value.data.size.toLong()
        }
        if (storedBytes + bytes <= maximumBytes) {
            entries[key] = file
            storedBytes += bytes
        }
    }

    private fun releaseWaiter(key: AttachmentDownloadKey, pending: Pending) {
        val orphan = synchronized(lock) {
            val current = inFlight[key]
            if (current?.token !== pending.token) return@synchronized null
            current.waiters = (current.waiters - 1).coerceAtLeast(0)
            if (current.waiters == 0) {
                inFlight.remove(key)
                current.deferred
            } else {
                null
            }
        }
        orphan?.cancel()
    }
}
